File: model/src/main/java/pl/agh/harmonytools/model/ChordComponent.java
package pl.agh.harmonytools.model;

import lombok.Getter;

import java.util.HashMap;

@Getter
public class ChordComponent {

    static final HashMap<Integer, Integer> baseComponentPitch = new HashMap<>();
    static {
        baseComponentPitch.put(1, 0);
        baseComponentPitch.put(2, 2);
        baseComponentPitch.put(3, 4);
        baseComponentPitch.put(4, 5);
        baseComponentPitch.put(5, 7);
        baseComponentPitch.put(6, 9);
        baseComponentPitch.put(7, 10);
        baseComponentPitch.put(8, 12);
        baseComponentPitch.put(9, 14);
    }

    private String chordComponentString;
    private Integer baseComponent;
    private long semitonesNumber;
    private boolean isDown;

    public ChordComponent(String chordComponentString, boolean isDown){
        this.chordComponentString = chordComponentString;
        this.baseComponent = Integer.parseInt(chordComponentString.substring(0,1));
        this.semitonesNumber = baseComponentPitch.get(baseComponent)
                + chordComponentString.codePoints().filter(ch -> ch == '<').count()
                - chordComponentString.codePoints().filter(ch -> ch == '>').count()
                + (isDown ? -1 : 0);
        this.isDown = isDown;
    }

    // should be moved from here
    public String toXmlString() {
        return this.chordComponentString.replace("<", "&lt;").replace(">", "&gt;");
    }
}

File: model/src/main/scala/pl/agh/harmonytools/model/chord/ChordComponent.scala
package pl.agh.harmonytools.model.chord

import scala.collection.immutable.HashMap

case class ChordComponent(
  chordComponentString: String,
  baseComponent: Int,
  semitonesNumber: Int,
  isDown: Boolean
) {
  override def toString: String = chordComponentString

  lazy val toXmlString: String =
    chordComponentString.replace("<", "&lt;").replace(">", "&gt;")
}

object ChordComponent {

  final val baseComponentPitch: HashMap[Int, Int] = HashMap(
    1 -> 0,
    2 -> 2,
    3 -> 4,
    4 -> 5,
    5 -> 7,
    6 -> 9,
    7 -> 10,
    8 -> 12,
    9 -> 14,
    10 -> 16,
    11 -> 17,
    12 -> 19,
    13 -> 21
  )

  def apply(chordComponentString: String, isDown: Boolean = false): ChordComponent = {
    val baseComponent   = {
      if (chordComponentString.head.isDigit) Integer.parseInt(chordComponentString.takeWhile(_.isDigit))
      else Integer.parseInt(chordComponentString.reverse.takeWhile(_.isDigit).reverse)
    }
    var semitonesNumber = baseComponentPitch.getOrElse(baseComponent, throw new IllegalArgumentException("Illegal baseComponent: " + baseComponent))
    semitonesNumber += chordComponentString.count(_ == '<')
    semitonesNumber -= chordComponentString.count(_ == '>')
    new ChordComponent(
      chordComponentString,
      baseComponent,
      semitonesNumber,
      isDown
    )
  }
}
